src/node/mod.rs

```diff
@@ -0,0 +1,410 @@
+        DrmNode::from_stat(stat)
+    }
+
+    /// Creates a DRM node from a file stat.
+    pub fn from_stat(stat: Stat) -> Result<DrmNode, CreateDrmNodeError> {
+        let dev = stat.st_rdev;
+        DrmNode::from_dev_id(dev)
+    }
+
+    /// Creates a DRM node from a [`dev_t`].
+    pub fn from_dev_id(dev: dev_t) -> Result<DrmNode, CreateDrmNodeError> {
+        if !is_device_drm(dev) {
+            return Err(CreateDrmNodeError::NotDrmNode);
+        }
+
+        // The type of the DRM node is determined by the minor number ranges:
+        //   0 -  63 -> Primary
+        //  64 - 127 -> Control
+        // 128 - 255 -> Render
+        let ty = match minor(dev) >> 6 {
+            0 => NodeType::Primary,
+            1 => NodeType::Control,
+            2 => NodeType::Render,
+            _ => return Err(CreateDrmNodeError::NotDrmNode),
+        };
+
+        Ok(DrmNode { dev, ty })
+    }
+
+    /// Returns the type of the DRM node.
+    pub fn ty(&self) -> NodeType {
+        self.ty
+    }
+
+    /// Returns the device_id of the underlying DRM node.
+    pub fn dev_id(&self) -> dev_t {
+        self.dev
+    }
+
+    /// Returns the path of the open device if possible.
+    pub fn dev_path(&self) -> Option<PathBuf> {
+        node_path(self, self.ty).ok()
+    }
+
+    /// Returns the path of the specified node type matching the device, if available.
+    pub fn dev_path_with_type(&self, ty: NodeType) -> Option<PathBuf> {
+        node_path(self, ty).ok()
+    }
+
+    /// Returns a new node of the specified node type matching the device, if available.
+    pub fn node_with_type(&self, ty: NodeType) -> Option<Result<DrmNode, CreateDrmNodeError>> {
+        self.dev_path_with_type(ty).map(DrmNode::from_path)
+    }
+
+    /// Returns the major device number of the DRM device.
+    pub fn major(&self) -> u32 {
+        major(self.dev_id())
+    }
+
+    /// Returns the minor device number of the DRM device.
+    pub fn minor(&self) -> u32 {
+        minor(self.dev_id())
+    }
+
+    /// Returns whether the DRM device has render nodes.
+    pub fn has_render(&self) -> bool {
+        #[cfg(target_os = "linux")]
+        {
+            node_path(self, NodeType::Render).is_ok()
+        }
+
+        // TODO: More robust checks on non-linux.
+
+        #[cfg(target_os = "freebsd")]
+        {
+            false
+        }
+
+        #[cfg(not(any(target_os = "linux", target_os = "freebsd")))]
+        {
+            false
+        }
+    }
+}
+
+impl Display for DrmNode {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}{}", self.ty.minor_name_prefix(), minor(self.dev_id()))
+    }
+}
+
+/// A type of node
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
+pub enum NodeType {
+    /// A primary node may be used to allocate buffers.
+    ///
+    /// If no other node is present, this may be used to post a buffer to an output with mode-setting.
```

Comment:
I've never fully understood this, but does this ultimately decide whether we can derive `control::Device` in addition to `Device`?

And as with my "sequence" Pull Request, do some things currently in `control` belong to the main device?

Reply:
I'd have to delegate this question to @i509VCB. I'm not familiar with these traits.

Reply:
The return of this value is probably a bit more dynamic than statically typed.

Also interesting that you now see control nodes in the DRM device folders?